src/pudl/load/sqlite.py

```diff
@@ -0,0 +1,43 @@
+"""Load PUDL data into an SQLite database."""
+
+import logging
+
+import sqlalchemy as sa
+
+from pudl.metadata.classes import Package, Resource
+from pudl.metadata.resources import RESOURCE_METADATA
+
+logger = logging.getLogger(__name__)
+
+
+def dfs_to_sqlite(dfs, engine, foreign_keys="OFF"):
+    """Load a dictionary of dataframes into the PUDL DB."""
+    # This magic makes SQLAlchemy tell SQLite to check foreign key constraints
+    # whenever we insert data into thd database, which it doesn't do by default
+    @sa.event.listens_for(sa.engine.Engine, "connect")
```

Comment:
If I understand correctly, this is activating a hook in SQLAlchemy more generally, that tells it to listen for any connection to an SQLite database, and when it hears one, prepend this `PRAGMA` statement which tells SQLite to check FK constraints. So it applies everywhere once this has been run, for the rest of the application's execution, unless it were to be deactivated similarly somewhere else.

Reply:
wow okay. but... ugh i have a hard tome understanding magic. is it activating when an engine in connected to? is that what the connect comes form?

Reply:
We've typically used the highest level abstraction of SQLAlchemy and let it and Pandas figure out what needs to be done, but under the hood there are a bunch of different steps involved in getting from `df.to_sql()` to an actual database table. Connections, cursors, sessions, grouping of insert statements into bulk transactions, etc. The `PRAGMA` commands in SQLite let you set some parameters that apply for everything that will happen within a particular session of use. If I understand correctlty, all this is doing is listening for when a new connection happens and running this `PRAGMA` command first, so that it applies to any new database operations. It listens for connections because the `PRAGMA` will remain in effect for everything within that connection. But it would need to be re-applied for a new connection.

So when we call `df.to_sql(engine)` pandas uses the (database connection) engine we've given it to fire up a new connection, and does a bunch of bulk insertions with `execute_many` (my god it's so much faster than it used to be) and then when it's all done, closes the connection. The engine persists, because it's kind of a long-term factory for creating new connections, not an individual persistent connection.

Reply:
This seemed to be the accepted way to tell SQLite to enforce foreign key constraints with SQLAlchemy. I think there are a bunch of other listening hooks for injecting other kinds of stuff at the appropriate point in SQL statements as needed too.